pkg/frontend/admin_openshiftversion_put.go

```diff
@@ -0,0 +1,93 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"encoding/json"
+	"net/http"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/api/admin"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+)
+
+func (f *frontend) putAdminOpenShiftVersion(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	converter := f.apis[admin.APIVersion].OpenShiftVersionConverter()
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	if len(body) == 0 || !json.Valid(body) {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized.")
+		return
+	}
+
+	var version admin.OpenShiftVersion
+
+	err := json.Unmarshal(body, &version)
+	if err != nil {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content could not be deserialized: "+err.Error())
+		return
+	}
+
+	docs, err := f.dbOpenShiftVersions.ListAll(ctx)
+	if err != nil {
+		api.WriteError(w, http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "Internal server error.")
+		return
+	}
+
+	var foundDoc *api.OpenShiftVersionDocument
+	var returnDoc *api.OpenShiftVersionDocument
+
+	if docs != nil {
+		for _, doc := range docs.OpenShiftVersionDocuments {
+			if doc.OpenShiftVersion.Version == version.Version {
+				foundDoc = doc
+				break
+			}
+		}
+	}
+
+	isCreate := false
+	if foundDoc != nil {
+		returnDoc, err = f.dbOpenShiftVersions.Patch(ctx, foundDoc.ID, func(osvd *api.OpenShiftVersionDocument) error {
+			osvd.OpenShiftVersion.OpenShiftPullspec = version.OpenShiftPullspec
+			osvd.OpenShiftVersion.InstallerPullspec = version.InstallerPullspec
+			osvd.OpenShiftVersion.Enabled = version.Enabled
+			return nil
+		})
+		if err != nil {
+			api.WriteError(w, http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "Internal server error.")
+			return
+		}
+	} else {
+		isCreate = true
+		returnDoc, err = f.dbOpenShiftVersions.Create(ctx, &api.OpenShiftVersionDocument{
```

Comment:
There needs to be some input validate before this line, because as it stands we are able to PUT an empty `{}` and that results in an "OpenShiftVersion" with no `Version` that can never be removed. It also results in a panic in `getAdminOpenShiftVersions`, because the sort is not checking if the Version field is not nil.

We should make all Version, OpenShiftPullspec, and InstallerPullspec required during creation, otherwise we are able to PUT a version that has no image and/or installer pull spec, which is not usable. That would require any calling code to perform the validation in N places, when we should really do it here.

For an update PUT (PATCH?), we could either require all or more concisely allow individual field values to be specified for OpenShiftPullspec, InstallerPullspec, and/or Enabled and only modify the fields that have been specified. As it is now, attempting to enable a version with `PUT {"version":"4.10.20","enabled":true}` results in the OpenShiftPullspec and InstallerPullspec values being removed.